tests/Integration/Behaviour/Features/Context/Domain/CustomerFeatureContext.php

```diff
@@ -0,0 +1,106 @@
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace Tests\Integration\Behaviour\Features\Context\Domain;
+
+use Customer;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Command\SetPrivateNoteAboutCustomerCommand;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Command\SetRequiredFieldsForCustomerCommand;
+use PrestaShop\PrestaShop\Core\Domain\Customer\Query\GetRequiredFieldsForCustomer;
+use RuntimeException;
+
+class CustomerFeatureContext extends AbstractDomainFeatureContext
+{
+    /**
+     * @Given /^"(Partner offers)" is "(required|not required)"$/
+     * @Then /^"(Partner offers)" should be "(required|not required)"$/
+     */
+    public function validateRequiredFieldStatus($requiredField, $status)
+    {
+        $requiredFieldName = $this->getRequiredFieldName($requiredField);
+        $isRequired = $status === 'required';
+
+        $requiredFields = $this->getQueryBus()->handle(new GetRequiredFieldsForCustomer());
+
+        if ($isRequired && !in_array($requiredFieldName, $requiredFields, true)) {
+            throw new RuntimeException(sprintf('"%s" was expected to be required customer field.', $requiredField));
+        }
+
+        if (!$isRequired && in_array($requiredFieldName, $requiredFields, true)) {
+            throw new RuntimeException(sprintf('"%s" was not expected to be required customer field.', $requiredField));
+        }
+    }
+
+    /**
+     * @Given /^I specify "(Partner offers)" to be "(required|not required)"$/
+     */
+    public function specifyRequiredField($requiredField, $status)
+    {
+        $requiredFieldName = $this->getRequiredFieldName($requiredField);
+        $isRequired = $status === 'required';
+
+        $requiredFields = $isRequired ? [$requiredFieldName] : [];
+
+        $this->getSharedStorage()->set('customer_required_fields', $requiredFields);
+    }
+
+    /**
+     * @When I save required fields for customer
+     */
+    public function saveSpecifiedRequiredFields()
+    {
+        $requiredFields = $this->getSharedStorage()->get('customer_required_fields');
+
+        $this->getCommandBus()->handle(new SetRequiredFieldsForCustomerCommand($requiredFields));
+    }
+
+    /**
+     * @When I set :privateNote private note about customer :reference
```

Comment:
```suggestion
     * @When I write a private note for customer :reference with the content :content
```
Let's aim for scenarios that sound, as much as possible, feature-oriented 😉 

I know this is hard because the BO is basically a big CRUD. But we must try to avoid using generic CRUD-related words like "save", "set", "edit", "modify".

Example: when we migrate Customer Service page, instead of writing "I save a message for customer X" we can use "I answer to customer X message" 😉 this sentence carries the meaning, the goal, rather than the action.

Reply:
> words like "save", "set", "edit", "modify".

I did use "set". :smile: 

Actually, I've had discussion quite some time ago with Pablo about this private note and we settled with "set private note **about** customer". Using "save/set private note **for** customer" would be wrong, because it is not visible for customer itself, only for BO admins, that's why "**about** customer" was kept, as it is private note. This is very feature-oriented imo. :)

Reply:
OK for the "about", what about the "write" then 😄 ?

Reply:
Hmm, the use case is to set private note, the command says "set", why add new wording. :thinking: